### PR TITLE
Fix `enq_work` behavior when single-threaded

### LINKED ISSUE #47756: Tasks launched on foreign threads are getting stickied to them
We are probably noticing the absence of a threadpool here, and so making the new Task sticky, but there is the main task (1) which is also considered a background worker thread:
```julia
$ ./julia -t1 -q
julia> threadwork = @cfunction ()->(Threads.@spawn println(Threads.threadid(), " ", current_task().rngState0, " ", current_task().rngState1, " ", current_task().rngState2, " ", current_task().rngState3); Cint(0)) Cint ()
Ptr{Nothing} @0x000000010a7341f0

julia> tid = UInt[0]; @ccall pthread_create(tid::Ptr{UInt}, C_NULL::Ptr{Cvoid}, threadwork::Ptr{Cvoid}, C_NULL::Ptr{Cvoid})::Cint; @ccall pthread_join(tid[]::UInt, C_NULL::Ptr{Cvoid})::Cint
0

julia> Base.Workqueues
2-element Vector{Base.IntrusiveLinkedListSynchronized{Task}}:
 Base.IntrusiveLinkedListSynchronized{Task}(Base.IntrusiveLinkedList{Task}(nothing, nothing), Base.Threads.SpinLock(0))
 Base.IntrusiveLinkedListSynchronized{Task}(Base.IntrusiveLinkedList{Task}(Task (runnable) @0x0000000106fc8160, Task (runnable) @0x0000000106fc8160), Base.Threads.SpinLock(0))
```
This snippet works if we launch with `-t2` or greater

### LINKED ISSUE #48644: Cannot spawn task on interactive thread after starting Julia 1.9.0-beta4 with `-t1,1`
The CLI help for `-t` states:
```
-t, --threads {auto|N[,auto|M]}
       Enable N[+M] threads; N threads are assigned to the `default`
       threadpool, and if M is specified, M threads are assigned to the
       `interactive` threadpool...
```
So I would expect to be able to spawn a task on the interactive thread when starting Julia with `-t1,1` (one default and one interactive thread). But this fails on Julia 1.9.0-beta4:

```
$ julia +beta -q -t1,1

julia> fetch(Threads.@spawn :interactive Threads.threadpool())
:default
```
It works when starting Julia with `-t2,1` (two default and one interactive thread):
```
$ julia +beta -q -t2,1

julia> fetch(Threads.@spawn :interactive Threads.threadpool())
:interactive
```
I think this is a bug.
```julia
julia> versioninfo()
Julia Version 1.9.0-beta4
Commit b75ddb787ff (2023-02-07 21:53 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake)
  Threads: 2 on 8 virtual cores
```

### PR DESCRIPTION
**Pull request by kpamnany**
If there's only one thread in the task's preferred thread pool, use that thread's work queue.

Fixes #48644.
Fixes https://github.com/JuliaLang/julia/issues/47756

```bash
$ ./julia -t 1,1
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.0-DEV.628 (2023-02-17)
 _/ |\__'_|_|_|\__'_|  |  kp/fix-48644/2420ad1286 (fork: 1 commits, 0 days)
|__/                   |

julia> fetch(Threads.@spawn :interactive Threads.threadpool())
:interactive

julia>
```

### CODE REVIEW

**Review by vtjnash — COMMENTED** [review]
I don't think this is quite right either. I think the issue is that where we set `tid = Threads.threadid()`, it should first check if `t` has a preferred thread pool instead, instead of pinning it to the current thread.

A similar change will be needed in the other enq_work-like functions (e.g. wait2)

**Comment by kpamnany** [comment]
> I think the issue is that where we set `tid = Threads.threadid()`, it should first check if `t` has a preferred thread pool instead, instead of pinning it to the current thread.

My understanding of the logic is as follows.

If `maxthreadid()` is 1, the multiqueue is never used so all tasks go to the single thread's work queue whether they are sticky or not.

Otherwise, if `t` is sticky but has no `tid`, it is stuck to the current thread and the current task is also made sticky.

If `t` has a thread pool, we can stick it to one of that pool's threads, but what do we do with the current task (which is already running on the current thread)?

**Comment by vtjnash** [comment]
maxthreadid is dynamic, so should not be used for static scheduling decisions is the main challenge

**Comment by kpamnany** [comment]
This particular case is not a static scheduling decision though? If `maxthreadid()` were to increase from 1, the task's stickiness starts to matter, so subsequent scheduling decisions will be different, as they should be. Not seeing the problem.

Currently, the only way to create a task on a different thread pool is with `@spawn`, and such tasks are not sticky. So it should not be possible for `t` to be sticky *and* have a different thread pool from the caller.

**Comment by simsurace** [comment]
Is the issue that the threadpool size cannot be determined for the interactive thread pool if this pool does not exist? If an interactive threadpool is not available, what should happen when spawning a task on it? Should this error or simply revert to the default pool?

File: base/task.jl [hunk 1]
```diff
@@ -767,7 +767,7 @@ end
 
 function enq_work(t::Task)
     (t._state === task_state_runnable && t.queue === nothing) || error("schedule: Task not runnable")
-    if t.sticky || Threads.threadpoolsize() == 1
+    if t.sticky || Threads.maxthreadid() == 1
```
Comment (vtjnash) [inline base/task.jl:770]:
```suggestion
    if !t.sticky
        tpid = max(threadpoolid(t), 1)
        if tpid == 1 && Threads.threadpoolsize(tpid) == 1
            ccall(:jl_set_task_tid, Cint, (Any, Cint), t, 1)
            t.sticky = true
        end
    end
    if t.sticky
```
I think this was trying to check specifically the size of the first threadpool, and check if `t` is meant to run in the first threadpool, and if both are 1, then to add it to the sticky queue for the first threadpool.

File: base/task.jl [hunk 2]
```diff
@@ -767,7 +767,7 @@ end
...
         tid = Threads.threadid(t)
         if tid == 0
```
Comment (vtjnash) [inline base/task.jl:772]: Alternatively, simply this:
```suggestion
        tid = t.sticky ? Threads.threadid(t) : 1
        if tid == 0
```

Comment (vtjnash) [reply base/task.jl:772]: Note that the same change may be need in other `_wait2` functions in this file

File: base/task.jl (hunk 1 above)
Comment (kpamnany) [reply base/task.jl]: I reworked this to check for the size of the task's preferred thread pool, whichever it is.

File: base/task.jl (hunk 2 above)
Comment (kpamnany) [reply base/task.jl:774]: I don't think so; can you verify?

**Comment by kpamnany** [comment]
> Is the issue that the threadpool size cannot be determined for the interactive thread pool if this pool does not exist? If an interactive threadpool is not available, what should happen when spawning a task on it? Should this error or simply revert to the default pool?

Sorry, forgot to answer this. The recent commits should make it clearer, but basically if there's only one thread in whichever thread pool a task  is to run in, we want to behave as though that task is sticky (i.e. use the thread's work queue rather than the multiqueue) as a small optimization.

Your other question should also be answered by one of the commits here -- if there is no interactive thread pool, we simply revert to the default thread pool.

**Comment by kpamnany** [comment]
Is this done then @vtjnash?

**Pull request merged**